selfdrive/locationd/locationd.cc

```diff
@@ -0,0 +1,378 @@
+  initMeasurement(fix.initVelocityNED(), ned_vel, nans, true);
+  initMeasurement(fix.initVelocityDevice(), vel_device, vel_device_std, true);
+  initMeasurement(fix.initAccelerationDevice(), accDevice, accDeviceErr, true);
+  initMeasurement(fix.initOrientationECEF(), orientation_ecef, orientation_ecef_std, true);
+  initMeasurement(fix.initCalibratedOrientationECEF(), calibrated_orientation_ecef, nans, this->calibrated);
+  initMeasurement(fix.initOrientationNED(), orientation_ned, nans, true);
+  initMeasurement(fix.initAngularVelocityDevice(), angVelocityDevice, angVelocityDeviceErr, true);
+  initMeasurement(fix.initVelocityCalibrated(), vel_calib, vel_calib_std, this->calibrated);
+  initMeasurement(fix.initAngularVelocityCalibrated(), ang_vel_calib, ang_vel_calib_std, this->calibrated);
+  initMeasurement(fix.initAccelerationCalibrated(), acc_calib, acc_calib_std, this->calibrated);
+
+  double old_mean = 0.0, new_mean = 0.0;
+  int i = 0;
+  for (double x : this->posenet_stds) {
+    if (i < POSENET_STD_HIST_HALF) {
+      old_mean += x;
+    } else {
+      new_mean += x;
+    }
+    i++;
+  }
+  old_mean /= POSENET_STD_HIST_HALF;
+  new_mean /= POSENET_STD_HIST_HALF;
+  // experimentally found these values, no false positives in 20k minutes of driving
+  bool std_spike = (new_mean / old_mean > 4.0 && new_mean > 7.0);
+
+  fix.setPosenetOK(!(std_spike && this->car_speed > 5.0));
+  fix.setDeviceStable(!this->device_fell);
+  this->device_fell = false;
+
+  //fix.setGpsWeek(this->time.week);
+  //fix.setGpsTimeOfWeek(this->time.tow);
+  fix.setUnixTimestampMillis(this->unix_timestamp_millis);
+
+  if (fix_ecef_std.norm() < 50.0 && this->calibrated) {
+    fix.setStatus(cereal::LiveLocationKalman::Status::VALID);
+  } else if (fix_ecef_std.norm() < 50.0) {
+    fix.setStatus(cereal::LiveLocationKalman::Status::UNCALIBRATED);
+  } else {
+    fix.setStatus(cereal::LiveLocationKalman::Status::UNINITIALIZED);
+  }
+}
+
+VectorXd Localizer::getPositionGeodetic() {
+  VectorXd fix_ecef = this->kf->get_x().segment<STATE_ECEF_POS_LEN>(STATE_ECEF_POS_START);
+  ECEF fix_ecef_ecef = { .x = fix_ecef(0), .y = fix_ecef(1), .z = fix_ecef(2) };
+  Geodetic fix_pos_geo = ecef2geodetic(fix_ecef_ecef);
+  return Vector3d(fix_pos_geo.lat, fix_pos_geo.lon, fix_pos_geo.alt);
+}
+
+void Localizer::handle_sensors(double current_time, const capnp::List<cereal::SensorEventData, capnp::Kind::STRUCT>::Reader& log) {
+  // TODO does not yet account for double sensor readings in the log
+  for (int i = 0; i < log.size(); i++) {
+    const cereal::SensorEventData::Reader& sensor_reading = log[i];
+    double sensor_time = 1e-9 * sensor_reading.getTimestamp();
+    // TODO: handle messages from two IMUs at the same time
+    if (sensor_reading.getSource() == cereal::SensorEventData::SensorSource::LSM6DS3) {
+      continue;
+    }
+
+    // Gyro Uncalibrated
+    if (sensor_reading.getSensor() == 5 && sensor_reading.getType() == 16) {
```

Comment:
```suggestion
    if (sensor_reading.getSensor() == cereal::SensorEventData::SensorSource::LSM6DS3 && sensor_reading.getType() == SENSOR_TYPE_GYROSCOPE_UNCALIBRATED) {
```

Reply:
This is not correct. The first field is an integer and comes from some android sensor id.